From https://github.com/oracle/solaris-userland

Allow for os.posix_spawn call with env set to None, which makes it inherit
the environment from the current process (similar to execv).

With this change, subprocess using posix_spawn can behave similarly
to the standard fork exec path.

This was offered and merged upstream; it will be available in Python 3.13:
https://github.com/python/cpython/issues/113119
https://github.com/python/cpython/pull/113120

diff -wpruN --no-dereference '--exclude=*.orig' a~/Lib/subprocess.py a/Lib/subprocess.py
--- a~/Lib/subprocess.py	1970-01-01 00:00:00
+++ a/Lib/subprocess.py	1970-01-01 00:00:00
@@ -1755,9 +1755,6 @@ class Popen:
                          c2pread, c2pwrite,
                          errread, errwrite):
             """Execute program using os.posix_spawn()."""
-            if env is None:
-                env = os.environ
-
             kwargs = {}
             if restore_signals:
                 # See _Py_RestoreSignals() in Python/pylifecycle.c
diff -wpruN --no-dereference '--exclude=*.orig' a~/Modules/posixmodule.c a/Modules/posixmodule.c
--- a~/Modules/posixmodule.c	1970-01-01 00:00:00
+++ a/Modules/posixmodule.c	1970-01-01 00:00:00
@@ -7061,9 +7061,9 @@ py_posix_spawn(int use_posix_spawnp, PyO
         return NULL;
     }
 
-    if (!PyMapping_Check(env)) {
+    if (!PyMapping_Check(env) && env != Py_None) {
         PyErr_Format(PyExc_TypeError,
-                     "%s: environment must be a mapping object", func_name);
+                     "%s: environment must be a mapping object or None", func_name);
         goto exit;
     }
 
@@ -7077,10 +7077,14 @@ py_posix_spawn(int use_posix_spawnp, PyO
         goto exit;
     }
 
+    if (env == Py_None) {
+        envlist = environ;
+    } else {
     envlist = parse_envlist(env, &envc);
     if (envlist == NULL) {
         goto exit;
     }
+    }
 
     if (file_actions != NULL && file_actions != Py_None) {
         /* There is a bug in old versions of glibc that makes some of the
@@ -7142,7 +7146,7 @@ exit:
     if (attrp) {
         (void)posix_spawnattr_destroy(attrp);
     }
-    if (envlist) {
+    if (envlist && envlist != environ) {
         free_string_array(envlist, envc);
     }
     if (argvlist) {
